src/lib/blockchain_snark/blockchain_snark_state.ml

```diff
@@ -74,36 +74,46 @@ module Make_update (T : Transaction_snark.Verification.S) = struct
         and supply_increase_is_zero =
           Currency.Amount.(equal_var supply_increase (var_of_t zero))
         in
-        let%bind nothing_changed =
-          Boolean.(ledger_hash_didn't_change && supply_increase_is_zero)
-        in
-        let%bind new_pending_coinbase_hash, deleted_stack =
+        let%bind new_pending_coinbase_hash, deleted_stack, no_coinbases_popped
+            =
           let%bind root_after_delete, deleted_stack =
             Pending_coinbase.Checked.pop_coinbases prev_pending_coinbase_root
               ~proof_emitted:(Boolean.not ledger_hash_didn't_change)
           in
+          let%bind prev_state_body_hash =
+            Protocol_state.(Body.hash_checked (body previous_state))
+          in
+          let%bind no_coinbases_popped =
+            let%bind check =
+              Pending_coinbase.Hash.equal_var root_after_delete
+                prev_pending_coinbase_root
+            in
+            Boolean.if_ ledger_hash_didn't_change ~then_:check
```

Comment:
I don't totally understand why this should be conditional. Could you add a comment explaining?

Reply:
If there's no new proof then `pop_coinbases` should be a no-op. I'll put a comment there